Add LollipopGraph

marinaanagno opened the pull request.

samanthaharper commented on doc/examples.xml:
```diff
@@ -309,3 +309,35 @@ gap> GeneralisedPetersenGraph(IsMutableDigraph, 9, 4);
+    <P/>
+
+<Example><![CDATA[
+gap> LollipopGraph(5, 4);
```
Make sure the functions returns an immutable digraph, per conversation with James.

samanthaharper commented on gap/examples.gi:
```diff
@@ -370,3 +370,13 @@ GeneralisedPetersenGraphCons);
+  local D;
+  D := DigraphDisjointUnion(CompleteDigraph(IsMutableDigraph, m),
+  DigraphSymmetricClosure(ChainDigraph(IsMutableDigraph, n)));
+  DigraphAddEdges(D, [[m, m + 1], [m + 1, m]]);
```
`MakeImmutable(D)`

samanthaharper reviewed (commented):
I think this is perfect, except that it should return an immutable digraph, so you need to update that and update the documentation.

alex-konovalov commented on doc/examples.xml:
```diff
@@ -309,3 +309,35 @@ gap> GeneralisedPetersenGraph(IsMutableDigraph, 9, 4);
+  <Oper Name="LollipopGraph" Arg="[filt, ]n, k"/>
+  <Returns>A digraph</Returns>
+  <Description>
+    If <A>m</A> and <A>n</A> are positive integers and <A>k</A>, 
```
What's `m`? I can't see it in the list of arguments

alex-konovalov replied:
Also, `k` is not explained. Looks like names mismatch with the actual code.

marinaanagno replied:
@alex-konovalov thanks for this! Apologies for the confusion! :)

alex-konovalov commented on doc/examples.xml:
```diff
@@ -309,3 +309,35 @@ gap> GeneralisedPetersenGraph(IsMutableDigraph, 9, 4);
+    path digraph on  <A>n</A> vertices connected by a bridge.
+
+    From
+    <URL>https://en.wikipedia.org/wiki/Lollipop_graph</URL>:
```
Why there is a colon after the URL - is there a text after which is missing?

marinaanagno replied:
Thanks @alex-konovalov probably a typo, fixed this!

wilfwilson commented:
You can see, for instance here: https://app.codecov.io/gh/digraphs/Digraphs/compare/417/diff, the bits of your added/changed code that don't have code coverage (you get there by following the "Details" link next to the failing `codecov/patch` PR check, and then following links...)

wilfwilson commented on gap/examples.gi:
```diff
@@ -370,3 +370,33 @@ GeneralisedPetersenGraphCons);
+  return D;
+end);
+
+InstallMethod(LollipopGraph, "for a function and two pos int",
```
This string does not match the arguments.

wilfwilson commented on doc/examples.xml:
```diff
@@ -309,3 +309,35 @@ gap> GeneralisedPetersenGraph(IsMutableDigraph, 9, 4);
+  <Returns>A digraph</Returns>
+  <Description>
+    If <A>m</A> and <A>n</A> are positive integers, 
+    then this operation returns the <A>(m, n)</A>-lollipop graph
+    that consists of a complete digraph on <A>m</A> vertices and a 
+    path digraph on  <A>n</A> vertices connected by a bridge.
+
+    From
+    <URL>https://en.wikipedia.org/wiki/Lollipop_graph</URL>
+    <P/>
```
I think this is a better way of putting it: (in particular, in Digraphs we say "chain digraph" rather than "path digraph".
```suggestion
    then this operation returns the <E>(<A>m</A>, <A>n</A>)-lollipop graph</E>.
    As defined at <URL>https://en.wikipedia.org/wiki/Lollipop_graph</URL>,
    this consists of a complete digraph on <A>m</A> vertices and the
    symmetric closure of a chain digraph on <A>n</A> vertices,
    connected by a bridge. <P/>
```
It is also undefined what a 'bridge' is. I think it's okay to use the word bridge initially. But afterwards, in any case, I think we need another paragraph to explicitly say what the vertices and edges are, similar to what I suggested in https://github.com/digraphs/Digraphs/pull/423/files#r592305981

In particular, in my opinion, you should explicitly note that:
* the complete digraph (the 'body' of the lollipop) is on the vertices `[1 .. m]`.
* the 'stick' of the lollipop is on the vertices `[m+1 .. m+n]`.
* The two parts are joined by the edge `[m, m + 1]` and its reverse.

wilfwilson commented on gap/examples.gi:
```diff
@@ -370,3 +370,33 @@ GeneralisedPetersenGraphCons);
+  local D;
+  D := MakeImmutable(LollipopGraphCons(IsMutableDigraph, m, n));
+  SetIsMultiDigraph(D, false);
+  SetIsSymmetricDigraph(D, true);
```
A few extra properties/attributes you could set:
```suggestion
  SetIsSymmetricDigraph(D, true);
  SetIsConnectedDigraph(D, true);
  SetDigraphHasLoops(D, false);
  SetChromaticNumber(D, Maximum(2, m));
  SetCliqueNumber(D, Maximum(2, m));
  if m <= 2 then
    SetDigraphUndirectedGirth(D, infinity);
  else
    SetDigraphUndirectedGirth(D, 3);
  fi;
```

wilfwilson reviewed (changes requested):
I think this is looking good! As with the other PRs, there will unfortunately be merge conflicts to resolve when you next update with the `master` branch. Let me know if that's trouble.

Otherwise, I have some requests about the documentation, and some suggestions for further properties to set about the lollipop graph.

wilfwilson reviewed (dismissed):
I've fixed some things that got broken by the merge conflicts, and also I addressed my remaining requested changes from https://github.com/digraphs/Digraphs/pull/417#discussion_r592412575 which you had already marked as resolved 😉 . I think this can be merged if the tests pass for a final time.

marinaanagno commented:
> I've fixed some things that got broken by the merge conflicts, and also I addressed my remaining requested changes from [#417 (comment)](https://github.com/digraphs/Digraphs/pull/417#discussion_r592412575) which you had already marked as resolved 😉 . I think this can be merged if the tests pass for a final time.

Ooops I'm so sorry! I don't know how I did this! Apologies! :)
Thank you so much for all this help!
